File: packages/barcode-scanning/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/barcodescanning/BarcodeScannerHelper.java
/**
 * Copyright (c) 2023 Robin Genz
 */
package io.capawesome.capacitorjs.plugins.mlkit.barcodescanning;

import android.graphics.Point;
import android.util.DisplayMetrics;
import android.util.Size;
import android.view.Display;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.google.mlkit.vision.barcode.common.Barcode;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONException;

public class BarcodeScannerHelper {

    public static JSObject createBarcodeResultForBarcode(@NonNull Barcode barcode, @Nullable Point imageSize, @Nullable Point screenSize) {
        Point[] cornerPoints = barcode.getCornerPoints();
        JSArray cornerPointsResult = new JSArray();
        if (cornerPoints != null && imageSize != null && screenSize != null) {
            Point[] normalizedCornerPoints = normalizeCornerPoints(cornerPoints, imageSize, screenSize);
            for (int i = 0; i < normalizedCornerPoints.length; i++) {
                JSArray cornerPointResult = new JSArray();
                cornerPointResult.put(normalizedCornerPoints[i].x);
                cornerPointResult.put(normalizedCornerPoints[i].y);
                cornerPointsResult.put(cornerPointResult);
            }
        } else if (cornerPoints != null && screenSize == null) {
            for (Point cornerPoint : cornerPoints) {
                JSArray cornerPointResult = new JSArray();
                cornerPointResult.put(cornerPoint.x);
                cornerPointResult.put(cornerPoint.y);
                cornerPointsResult.put(cornerPointResult);
            }
        }

        JSObject result = new JSObject();
        result.put("bytes", convertByteArrayToJsonArray(barcode.getRawBytes()));
        if (cornerPoints != null) {
            result.put("cornerPoints", cornerPointsResult);
        }
        if (barcode.getCalendarEvent() != null) {
            result.put("calendarEvent", extractCalendarEventProperties(barcode.getCalendarEvent()));
        }
        if (barcode.getContactInfo() != null) {
            result.put("contactInfo", extractContactInfoProperties(barcode.getContactInfo()));
        }
        result.put("displayValue", barcode.getDisplayValue());
        if (barcode.getDriverLicense() != null) {
            result.put("driverLicense", extractDriverLicenseProperties(barcode.getDriverLicense()));
        }
        if (barcode.getEmail() != null) {
            result.put("email", extractEmailProperties(barcode.getEmail()));
        }
        result.put("format", convertBarcodeScannerFormatToString(barcode.getFormat()));
        if (barcode.getGeoPoint() != null) {
            result.put("geoPoint", extractGeoPointProperties(barcode.getGeoPoint()));
        }
        if (barcode.getPhone() != null) {
            result.put("phone", extractPhoneProperties(barcode.getPhone()));
        }
        result.put("rawValue", barcode.getRawValue() == null ? "" : barcode.getRawValue());
        if (barcode.getSms() != null) {
            result.put("sms", extractSmsProperties(barcode.getSms()));
        }
        if (barcode.getUrl() != null) {
            result.put("urlBookmark", extractUrlBookmark(barcode.getUrl()));
        }
        result.put("valueType", convertBarcodeValueTypeToString(barcode.getValueType()));
        if (barcode.getWifi() != null) {
            result.put("wifi", extractWifiProperties(barcode.getWifi()));
        }
        return result;
    }

    public static int[] convertStringsToBarcodeScannerFormats(String[] values) throws JSONException {
        int[] formats = new int[values.length];
        for (int i = 0; i < values.length; i++) {
            Integer format = convertStringToBarcodeScannerFormat(values[i]);
            if (format == null) {
                continue;
            }
            formats[i] = format;
        }
        return formats;
    }

    @Nullable
    public static Integer convertStringToBarcodeScannerFormat(String value) {
        switch (value) {
            case "AZTEC":
                return Barcode.FORMAT_AZTEC;
            case "CODABAR":
                return Barcode.FORMAT_CODABAR;
            case "CODE_39":
                return Barcode.FORMAT_CODE_39;
            case "CODE_93":
                return Barcode.FORMAT_CODE_93;
            case "CODE_128":
                return Barcode.FORMAT_CODE_128;
            case "DATA_MATRIX":
                return Barcode.FORMAT_DATA_MATRIX;
            case "EAN_8":
                return Barcode.FORMAT_EAN_8;
            case "EAN_13":
                return Barcode.FORMAT_EAN_13;
            case "ITF":
                return Barcode.FORMAT_ITF;
            case "PDF_417":
                return Barcode.FORMAT_PDF417;
            case "QR_CODE":
                return Barcode.FORMAT_QR_CODE;
            case "UPC_A":
                return Barcode.FORMAT_UPC_A;
            case "UPC_E":
                return Barcode.FORMAT_UPC_E;
            default:
                return null;
        }
    }

    @Nullable
    public static String convertBarcodeScannerFormatToString(int format) {
        switch (format) {
            case Barcode.FORMAT_AZTEC:
                return "AZTEC";
            case Barcode.FORMAT_CODABAR:
                return "CODABAR";
            case Barcode.FORMAT_CODE_39:
                return "CODE_39";
            case Barcode.FORMAT_CODE_93:
                return "CODE_93";
            case Barcode.FORMAT_CODE_128:
                return "CODE_128";
            case Barcode.FORMAT_DATA_MATRIX:
                return "DATA_MATRIX";
            case Barcode.FORMAT_EAN_8:
                return "EAN_8";
            case Barcode.FORMAT_EAN_13:
                return "EAN_13";
            case Barcode.FORMAT_ITF:
                return "ITF";
            case Barcode.FORMAT_PDF417:
                return "PDF_417";
            case Barcode.FORMAT_QR_CODE:
                return "QR_CODE";
            case Barcode.FORMAT_UPC_A:
                return "UPC_A";
            case Barcode.FORMAT_UPC_E:
                return "UPC_E";
            default:
                return null;
        }
    }

    @Nullable
    public static String convertBarcodeValueTypeToString(int format) {
        switch (format) {
            case Barcode.TYPE_CALENDAR_EVENT:
                return "CALENDAR_EVENT";
            case Barcode.TYPE_CONTACT_INFO:
                return "CONTACT_INFO";
            case Barcode.TYPE_DRIVER_LICENSE:
                return "DRIVERS_LICENSE";
            case Barcode.TYPE_EMAIL:
                return "EMAIL";
            case Barcode.TYPE_GEO:
                return "GEO";
            case Barcode.TYPE_ISBN:
                return "ISBN";
            case Barcode.TYPE_PHONE:
                return "PHONE";
            case Barcode.TYPE_PRODUCT:
                return "PRODUCT";
            case Barcode.TYPE_SMS:
                return "SMS";
            case Barcode.TYPE_TEXT:
                return "TEXT";
            case Barcode.TYPE_URL:
                return "URL";
            case Barcode.TYPE_WIFI:
                return "WIFI";
            default:
                return "UNKNOWN";
        }
    }

    @Nullable
    private static JSONArray convertByteArrayToJsonArray(@Nullable byte[] bytes) {
        if (bytes == null) {
            return null;
        }
        JSONArray ret = new JSONArray();
        for (byte _byte : bytes) {
            ret.put(_byte);
        }
        return ret;
    }

    private static String convertCalendarDateTimeToString(Barcode.CalendarDateTime dateTime) {
        int year = dateTime.getYear() == -1 ? 0 : dateTime.getYear();
        int month = dateTime.getMonth() == -1 ? 0 : dateTime.getMonth();
        int day = dateTime.getDay() == -1 ? 0 : dateTime.getDay();
        int hour = dateTime.getHours() == -1 ? 0 : dateTime.getHours();
        int minute = dateTime.getMinutes() == -1 ? 0 : dateTime.getMinutes();
        int second = dateTime.getSeconds() == -1 ? 0 : dateTime.getSeconds();
        boolean isUtc = dateTime.isUtc();
        return String.format("%04d-%02d-%02dT%02d:%02d:%02d%s", year, month, day, hour, minute, second, isUtc ? "Z" : "");
    }

    private static JSArray extractAddresses(List<Barcode.Address> addresses) {
        JSArray addressArray = new JSArray();
        for (Barcode.Address address : addresses) {
            JSObject addressObj = new JSObject();
            addressObj.put("addressLines", address.getAddressLines());
            addressObj.put("type", address.getType());
            addressArray.put(addressObj);
        }

        return addressArray;
    }

    private static JSObject extractCalendarEventProperties(Barcode.CalendarEvent calendarEvent) {
        JSObject result = new JSObject();

        result.put("description", calendarEvent.getDescription());
        result.put("end", convertCalendarDateTimeToString(calendarEvent.getEnd()));
        result.put("location", calendarEvent.getLocation());
        result.put("organizer", calendarEvent.getOrganizer());
        result.put("start", convertCalendarDateTimeToString(calendarEvent.getStart()));
        result.put("status", calendarEvent.getStatus());
        result.put("summary", calendarEvent.getSummary());

        return result;
    }

    private static JSObject extractContactInfoProperties(Barcode.ContactInfo contactInfo) {
        JSObject result = new JSObject();

        result.put("addresses", extractAddresses(contactInfo.getAddresses()));
        result.put("emails", extractEmails(contactInfo.getEmails()));
        result.put("name", extractPersonNameProperties(contactInfo.getName()));
        result.put("organization", contactInfo.getOrganization());
        result.put("phones", extractPhones(contactInfo.getPhones()));
        result.put("title", contactInfo.getTitle());
        result.put("urls", extractUrls(contactInfo.getUrls()));

        return result;
    }

    private static JSObject extractDriverLicenseProperties(Barcode.DriverLicense driverLicense) {
        JSObject result = new JSObject();

        result.put("addressCity", driverLicense.getAddressCity());
        result.put("addressState", driverLicense.getAddressState());
        result.put("addressStreet", driverLicense.getAddressStreet());
        result.put("addressZip", driverLicense.getAddressZip());
        result.put("birthDate", driverLicense.getBirthDate());
        result.put("documentType", driverLicense.getDocumentType());
        result.put("expiryDate", driverLicense.getExpiryDate());
        result.put("firstName", driverLicense.getFirstName());
        result.put("gender", driverLicense.getGender());
        result.put("issueDate", driverLicense.getIssueDate());
        result.put("issuingCountry", driverLicense.getIssuingCountry());
        result.put("lastName", driverLicense.getLastName());
        result.put("licenseNumber", driverLicense.getLicenseNumber());
        result.put("middleName", driverLicense.getMiddleName());

        return result;
    }

    private static JSObject extractEmailProperties(Barcode.Email email) {
        JSObject emailObj = new JSObject();
        emailObj.put("address", email.getAddress());
        emailObj.put("body", email.getBody());
        emailObj.put("subject", email.getSubject());
        emailObj.put("type", email.getType());

        return emailObj;
    }

    private static JSArray extractEmails(List<Barcode.Email> emails) {
        JSArray emailArray = new JSArray();
        for (Barcode.Email email : emails) {
            JSObject emailObj = extractEmailProperties(email);
            emailArray.put(emailObj);
        }

        return emailArray;
    }

    private static JSObject extractGeoPointProperties(Barcode.GeoPoint geoPoint) {
        JSObject geoPointObj = new JSObject();
        geoPointObj.put("latitude", geoPoint.getLat());
        geoPointObj.put("longitude", geoPoint.getLng());

        return geoPointObj;
    }

    private static JSObject extractPersonNameProperties(Barcode.PersonName personName) {
        JSObject result = new JSObject();

        result.put("first", personName.getFirst());
        result.put("formattedName", personName.getFormattedName());
        result.put("last", personName.getLast());
        result.put("middle", personName.getMiddle());
        result.put("prefix", personName.getPrefix());
        result.put("pronunciation", personName.getPronunciation());
        result.put("suffix", personName.getSuffix());

        return result;
    }

    private static JSObject extractPhoneProperties(Barcode.Phone phone) {
        JSObject result = new JSObject();
        result.put("number", phone.getNumber());
        result.put("type", phone.getType());

        return result;
    }

    private static JSArray extractPhones(List<Barcode.Phone> phones) {
        JSArray phoneArray = new JSArray();
        for (Barcode.Phone phone : phones) {
            JSObject phoneObj = extractPhoneProperties(phone);
            phoneArray.put(phoneObj);
        }

        return phoneArray;
    }

    private static JSObject extractSmsProperties(Barcode.Sms sms) {
        JSObject result = new JSObject();
        result.put("message", sms.getMessage());
        result.put("phoneNumber", sms.getPhoneNumber());
        return result;
    }

    private static JSObject extractUrlBookmark(Barcode.UrlBookmark urlBookmark) {
        JSObject result = new JSObject();
        result.put("title", urlBookmark.getTitle());
        result.put("url", urlBookmark.getUrl());
        return result;
    }

    private static JSArray extractUrls(List<String> urls) {
        JSArray urlArray = new JSArray();
        for (String url : urls) {
            urlArray.put(url);
        }

        return urlArray;
    }

    private static JSObject extractWifiProperties(Barcode.WiFi wifi) {
        JSObject result = new JSObject();
        result.put("encryptionType", wifi.getEncryptionType());
        result.put("password", wifi.getPassword());
        result.put("ssid", wifi.getSsid());

        return result;
    }

    private static Point[] normalizeCornerPoints(@NonNull Point[] cornerPoints, @NonNull Point imageSize, @NonNull Point screenSize) {
        // Log corner points
        // Logger.debug("Corner points: " + cornerPoints[0] + ", " + cornerPoints[1] + ", " + cornerPoints[2] + ", " + cornerPoints[3]);
        double screenWidth = screenSize.x;
        double screenHeight = screenSize.y;
        double imageWidth = imageSize.x;
        double imageHeight = imageSize.y;
        // Swap the image dimensions if the image is in landscape mode
        if (screenWidth > screenHeight) {
            imageWidth = imageSize.y;
            imageHeight = imageSize.x;
        }
        // Calculate the scale of the image
        double scale = Math.max(screenHeight / imageWidth, screenWidth / imageHeight);
        // Calculate the invisible area of the image
        double invisibleWidth = imageHeight * scale - screenWidth;
        double invisibleHeight = imageWidth * scale - screenHeight;
        Point[] normalizedCornerPoints = new Point[cornerPoints.length];
        for (int i = 0; i < cornerPoints.length; i++) {
            // Scale the points and move them to the center of the screen
            int x = (int) ((cornerPoints[i].x * scale) - (invisibleWidth / 2));
            int y = (int) ((cornerPoints[i].y * scale) - (invisibleHeight / 2));
            normalizedCornerPoints[i] = new Point(x, y);
        }
        // Log normalized corner points
        // Logger.debug("Normalized corner points: " + normalizedCornerPoints[0] + ", " + normalizedCornerPoints[1] + ", " + normalizedCornerPoints[2] + ", " + normalizedCornerPoints[3]);
        return normalizedCornerPoints;
    }

    public static Size convertIntegerToResolution(Integer resolution) {
        return switch (resolution) {
            case 0 -> new Size(640, 480);
            case 2 -> new Size(1920, 1080);
            case 3 -> new Size(3840, 2160);
            default -> new Size(1280, 720);
        };
    }
}

File: packages/barcode-scanning/ios/Plugin/BarcodeScannerHelper.swift
/**
 * Copyright (c) 2023 Robin Genz
 */
import Foundation
import Capacitor
import MLKitBarcodeScanning
import AVFoundation

// swiftlint:disable cyclomatic_complexity
public class BarcodeScannerHelper {
    // swiftlint:disable identifier_name
    public static func normalizeCornerPoints(cornerPoints: [NSValue], imageSize: CGSize, screenSize: CGSize, videoOrientation: AVCaptureVideoOrientation?) -> [NSValue] {
        // Log corner points
        // CAPLog.print("Corner points (\(cornerPoints[0].cgPointValue.x), \(cornerPoints[0].cgPointValue.y)), (\(cornerPoints[1].cgPointValue.x), \(cornerPoints[1].cgPointValue.y)), (\(cornerPoints[2].cgPointValue.x), \(cornerPoints[2].cgPointValue.y)), (\(cornerPoints[3].cgPointValue.x), \(cornerPoints[3].cgPointValue.y))")
        let screenWidth = screenSize.width
        let screenHeight = screenSize.height
        var imageWidth = imageSize.width
        var imageHeight = imageSize.height
        // Swap the image dimensions if the image is in landscape mode
        if screenWidth > screenHeight {
            imageWidth = imageSize.height
            imageHeight = imageSize.width
        }
        // Calculate the scale of the image
        let scale = max(screenHeight / CGFloat(imageWidth), screenWidth / CGFloat(imageHeight))
        // Calculate the invisible area of the image
        let invisibleWidth = imageHeight * scale - screenWidth
        let invisibleHeight = imageWidth * scale - screenHeight
        var normalizedCornerPoints = [NSValue]()
        for cornerPoint in cornerPoints {
            var x: CGFloat
            var y: CGFloat
            switch videoOrientation {
            case .portrait, .portraitUpsideDown:
                x = ((imageHeight - cornerPoint.cgPointValue.y) * scale) - (invisibleWidth / 2)
                y = (cornerPoint.cgPointValue.x * scale) - (invisibleHeight / 2)
            case .landscapeLeft:
                x = ((imageHeight - cornerPoint.cgPointValue.x) * scale) - (invisibleWidth / 2)
                y = ((imageWidth - cornerPoint.cgPointValue.y) * scale) - (invisibleHeight / 2)
            default:
                x = (cornerPoint.cgPointValue.x * scale) - (invisibleWidth / 2)
                y = (cornerPoint.cgPointValue.y * scale) - (invisibleHeight / 2)
            }
            let point = CGPoint(x: Int(x), y: Int(y))
            let value = NSValue(cgPoint: point)
            normalizedCornerPoints.append(value)
        }
        // If the video orientation is not in landscape right mode, the corner points need to be rotated
        switch videoOrientation {
        case .portrait:
            let lastNormalizedCornerPoints = normalizedCornerPoints.removeLast()
            normalizedCornerPoints.insert(lastNormalizedCornerPoints, at: 0)
        case .landscapeLeft:
            var lastNormalizedCornerPoints = normalizedCornerPoints.removeLast()
            normalizedCornerPoints.insert(lastNormalizedCornerPoints, at: 0)
            lastNormalizedCornerPoints = normalizedCornerPoints.removeLast()
            normalizedCornerPoints.insert(lastNormalizedCornerPoints, at: 0)
        default:
            break
        }
        // Log normalized corner points
        // CAPLog.print("Normalized corner points (\(normalizedCornerPoints[0].cgPointValue.x), \(normalizedCornerPoints[0].cgPointValue.y)), (\(normalizedCornerPoints[1].cgPointValue.x), \(normalizedCornerPoints[1].cgPointValue.y)), (\(normalizedCornerPoints[2].cgPointValue.x), \(normalizedCornerPoints[2].cgPointValue.y)), (\(normalizedCornerPoints[3].cgPointValue.x), \(normalizedCornerPoints[3].cgPointValue.y))")
        return normalizedCornerPoints
    }

    public static func createBarcodeResultForBarcode(_ barcode: Barcode, imageSize: CGSize?, videoOrientation: AVCaptureVideoOrientation?) -> JSObject {
        var cornerPointsResult = [[Int]]()
        if let cornerPoints = barcode.cornerPoints, let imageSize = imageSize {
            let screenSize = CGSize(width: UIScreen.main.bounds.width * UIScreen.main.scale, height: UIScreen.main.bounds.height * UIScreen.main.scale)
            let normalizedCornerPoints = normalizeCornerPoints(cornerPoints: cornerPoints, imageSize: imageSize, screenSize: screenSize, videoOrientation: videoOrientation)
            for cornerPoint in normalizedCornerPoints {
                var value = [Int]()
                value.append(Int(cornerPoint.cgPointValue.x))
                value.append(Int(cornerPoint.cgPointValue.y))
                cornerPointsResult.append(value)
            }
        } else if let cornerPoints = barcode.cornerPoints, videoOrientation == nil {
            for cornerPoint in cornerPoints {
                var value = [Int]()
                value.append(Int(cornerPoint.cgPointValue.x))
                value.append(Int(cornerPoint.cgPointValue.y))
                cornerPointsResult.append(value)
            }
        }

        var result = JSObject()
        if let rawData = barcode.rawData {
            result["bytes"] = convertDataToJsonArray(rawData)
        }
        result["cornerPoints"] = cornerPointsResult
        if let calendarEvent = barcode.calendarEvent {
            result["calendarEvent"] = extractCalendarEventProperties(calendarEvent)
        }
        if let contactInfo = barcode.contactInfo {
            result["contactInfo"] = extractContactInfoProperties(contactInfo)
        }
        result["displayValue"] = barcode.displayValue
        if let driverLicense = barcode.driverLicense {
            result["driverLicense"] = extractDriverLicenseProperties(driverLicense)
        }
        if let email = barcode.email {
            result["email"] = extractEmailProperties(email)
        }
        result["format"] = convertBarcodeScannerFormatToString(barcode.format)
        if let geoPoint = barcode.geoPoint {
            result["geoPoint"] = extractGeoPointProperties(geoPoint)
        }
        if let phone = barcode.phone {
            result["phone"] = extractPhoneProperties(phone)
        }
        result["rawValue"] = barcode.rawValue ?? ""
        if let sms = barcode.sms {
            result["sms"] = extractSmsProperties(sms)
        }
        if let url = barcode.url {
            result["url"] = extractUrlBookmark(url)
        }
        result["valueType"] = convertBarcodeValueTypeToString(barcode.valueType)
        if let wifi = barcode.wifi {
            result["wifi"] = extractWifiProperties(wifi)
        }
        return result
    }

    public static func convertStringsToBarcodeScannerFormats(_ values: [String]) -> [BarcodeFormat] {
        var formats: [BarcodeFormat] = []
        for value in values {
            let format = convertStringToBarcodeScannerFormat(value)
            guard let format = format else {
                continue
            }
            formats.append(format)
        }
        return formats
    }

    private static func convertStringToBarcodeScannerFormat(_ value: String) -> BarcodeFormat? {
        switch value {
        case "AZTEC":
            return BarcodeFormat.aztec
        case "CODABAR":
            return BarcodeFormat.codaBar
        case "CODE_39":
            return BarcodeFormat.code39
        case "CODE_93":
            return BarcodeFormat.code93
        case "CODE_128":
            return BarcodeFormat.code128
        case "DATA_MATRIX":
            return BarcodeFormat.dataMatrix
        case "EAN_8":
            return BarcodeFormat.EAN8
        case "EAN_13":
            return BarcodeFormat.EAN13
        case "ITF":
            return BarcodeFormat.ITF
        case "PDF_417":
            return BarcodeFormat.PDF417
        case "QR_CODE":
            return BarcodeFormat.qrCode
        case "UPC_A":
            return BarcodeFormat.UPCA
        case "UPC_E":
            return BarcodeFormat.UPCE
        default:
            return nil
        }
    }

    private static func convertBarcodeScannerFormatToString(_ format: BarcodeFormat) -> String? {
        switch format {
        case BarcodeFormat.aztec:
            return "AZTEC"
        case BarcodeFormat.codaBar:
            return "CODABAR"
        case BarcodeFormat.code39:
            return "CODE_39"
        case BarcodeFormat.code93:
            return "CODE_93"
        case BarcodeFormat.code128:
            return "CODE_128"
        case BarcodeFormat.dataMatrix:
            return "DATA_MATRIX"
        case BarcodeFormat.EAN8:
            return "EAN_8"
        case BarcodeFormat.EAN13:
            return "EAN_13"
        case BarcodeFormat.ITF:
            return "ITF"
        case BarcodeFormat.PDF417:
            return "PDF_417"
        case BarcodeFormat.qrCode:
            return "QR_CODE"
        case BarcodeFormat.UPCA:
            return "UPC_A"
        case BarcodeFormat.UPCE:
            return "UPC_E"
        default:
            return nil
        }
    }

    private static func convertBarcodeValueTypeToString(_ type: BarcodeValueType) -> String {
        switch type {
        case BarcodeValueType.calendarEvent:
            return "CALENDAR_EVENT"
        case BarcodeValueType.contactInfo:
            return "CONTACT_INFO"
        case BarcodeValueType.driversLicense:
            return "DRIVERS_LICENSE"
        case BarcodeValueType.email:
            return "EMAIL"
        case BarcodeValueType.geographicCoordinates:
            return "GEO"
        case BarcodeValueType.ISBN:
            return "ISBN"
        case BarcodeValueType.phone:
            return "PHONE"
        case BarcodeValueType.product:
            return "PRODUCT"
        case BarcodeValueType.SMS:
            return "SMS"
        case BarcodeValueType.text:
            return "TEXT"
        case BarcodeValueType.URL:
            return "URL"
        case BarcodeValueType.wiFi:
            return "WIFI"
        default:
            return "UNKNOWN"
        }
    }

    private static func convertDataToJsonArray(_ data: Data) -> [UInt8] {
        return [UInt8](data)
    }

    private static func extractAddressProperties(_ address: BarcodeAddress) -> JSObject {
        var result = JSObject()
        result["addressLines"] = address.addressLines
        result["type"] = address.type.rawValue
        return result
    }

    private static func extractCalendarEventProperties(_ calendarEvent: BarcodeCalendarEvent) -> JSObject {
        var result = JSObject()
        result["description"] = calendarEvent.eventDescription
        result["end"] = calendarEvent.end
        result["location"] = calendarEvent.location
        result["organizer"] = calendarEvent.organizer
        result["start"] = calendarEvent.start
        result["status"] = calendarEvent.status
        result["summary"] = calendarEvent.summary
        return result
    }

    private static func extractContactInfoProperties(_ contactInfo: BarcodeContactInfo) -> JSObject {
        var result = JSObject()
        result["addresses"] = (contactInfo.addresses ?? []).map { address in extractAddressProperties(address) }
        result["emails"] = (contactInfo.emails ?? []).map { email in extractEmailProperties(email) }
        result["name"] = extractPersonNameProperties(contactInfo.name)
        result["organization"] = contactInfo.organization
        result["phones"] = (contactInfo.phones ?? []).map { phone in
            extractPhoneProperties(phone)
        }
        result["title"] = contactInfo.jobTitle
        result["urls"] = JSArray(contactInfo.urls ?? [])
        return result
    }

    private static func extractDriverLicenseProperties(_ driverLicense: BarcodeDriverLicense) -> JSObject {
        var result = JSObject()
        result["addressCity"] = driverLicense.addressCity
        result["addressState"] = driverLicense.addressState
        result["addressStreet"] = driverLicense.addressStreet
        result["addressZip"] = driverLicense.addressZip
        result["birthDate"] = driverLicense.birthDate
        result["documentType"] = driverLicense.documentType
        result["expiryDate"] = driverLicense.expiryDate
        result["firstName"] = driverLicense.firstName
        result["gender"] = driverLicense.gender
        result["issuingDate"] = driverLicense.issuingDate
        result["issuingCountry"] = driverLicense.issuingCountry
        result["lastName"] = driverLicense.lastName
        result["licenseNumber"] = driverLicense.licenseNumber
        result["middleName"] = driverLicense.middleName
        return result
    }

    private static func extractEmailProperties(_ email: BarcodeEmail) -> JSObject {
        var result = JSObject()
        result["address"] = email.address
        result["body"] = email.body
        result["subject"] = email.subject
        result["type"] = email.type.rawValue
        return result
    }

    private static func extractGeoPointProperties(_ geoPoint: BarcodeGeoPoint) -> JSObject {
        var result = JSObject()
        result["latitude"] = geoPoint.latitude
        result["longitude"] = geoPoint.longitude
        return result
    }

    private static func extractPersonNameProperties(_ personName: BarcodePersonName?) -> JSObject {
        var result = JSObject()
        result["first"] = personName?.first
        result["formattedName"] = personName?.formattedName
        result["last"] = personName?.last
        result["middle"] = personName?.middle
        result["prefix"] = personName?.prefix
        result["pronunciation"] = personName?.pronunciation
        result["suffix"] = personName?.suffix
        return result
    }

    private static func extractPhoneProperties(_ phone: BarcodePhone) -> JSObject {
        var result = JSObject()
        result["number"] = phone.number
        result["type"] = phone.type.rawValue
        return result
    }

    private static func extractSmsProperties(_ sms: BarcodeSMS) -> JSObject {
        var result = JSObject()
        result["message"] = sms.message
        result["phoneNumber"] = sms.phoneNumber
        return result
    }

    private static func extractUrlBookmark(_ bookmark: BarcodeURLBookmark) -> JSObject {
        var result = JSObject()
        result["title"] = bookmark.title
        result["url"] = bookmark.url
        return result
    }

    private static func extractWifiProperties(_ wifi: BarcodeWifi) -> JSObject {
        var result = JSObject()
        result["encryptionType"] = wifi.type.rawValue
        result["ssid"] = wifi.ssid
        result["password"] = wifi.password
        return result
    }

    public static func convertIntToCapturePreset(_ resolution: Int) -> AVCaptureSession.Preset {
        switch resolution {
        case 0:
            return AVCaptureSession.Preset.vga640x480
        case 2:
            return AVCaptureSession.Preset.hd1920x1080
        case 3:
            return AVCaptureSession.Preset.hd4K3840x2160
        default:
            return AVCaptureSession.Preset.hd1280x720
        }
    }

}
